### PR TITLE
routeOnClick upgrade

### PR DESCRIPTION
**Pull request by sdeering**

### CODE REVIEW

File: test/karma/directives/routeOnClick.test.js [hunk 1]
```diff
@@ -22,13 +22,50 @@ describe('routeOnClick directive', function() {
...
+
+      inject(function ($rootScope, $compile, State) {
+        State.set('page', 2);
+        let element = $compile('<a route-on-click="paginationUrlWriter()">Link</a>')($rootScope.$new());
+        $rootScope.$digest();
+        expect(element.attr('href')).toBe('#/page/2');
+
+        State.set('page', 3);
+        $rootScope.$digest();
+        expect(element.attr('href')).toBe('#/page/3');
```
Comment (sdeering) [inline test/karma/directives/routeOnClick.test.js]: @lee-borsboom not sure why this test isn't working, it should be triggering the paginationUrlWriter() to return a different URL for the href but the $watch is not being fired. Any ideas?

File: test/karma/karma.conf.js [hunk 2]
```diff
@@ -58,7 +58,7 @@ module.exports = function (config) {
 
     browsers: ['Chrome'],
 
-    reporters: ['spec'],
+    // reporters: ['spec'],
```
Comment (sdeering) [inline test/karma/karma.conf.js]: temp while testing will revert once its working

File: src/directives/routeOnClick.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ function routeOnClickFactory (Route, $location, $window, $timeout) {
       const LEFT_BUTTON = 0;
       const MIDDLE_BUTTON = 1;
 
-      if (element.tagName === 'A') {
+      if (element.tagName === 'A' || element[0].tagName === 'A') {
```
Comment (sdeering) [inline src/directives/routeOnClick.js]: this one is also interesting, it looks like the tests require it, BUT does the directive actually need it?

File: package.json [hunk 4]
```diff
@@ -34,11 +34,11 @@
     "grunt-contrib-watch": "^1.0.0",
     "grunt-eslint": "^19.0.0",
     "grunt-express-server": "^0.5.1",
-    "grunt-karma": "^0.12.1",
+    "grunt-karma": "^1.0.0",
     "grunt-ng-annotate": "^3.0.0",
     "jasmine-core": "^2.3.4",
     "jest": "^19.0.0",
-    "karma": "^0.13.19",
+    "karma": "^1.7.1",
```
Comment (sdeering) [inline package.json]: just a bump can revert if you want, tests still work

File: src/directives/routeOnClick.js [hunk 5]
```diff
@@ -54,16 +54,22 @@ function routeOnClickFactory (Route, $location, $window, $timeout) {
...
-        return scope.$watch(iAttrs.routeOnClick, () => {
-          return element.attr('href', getUrl());
+        return scope.$watch(attrs.routeOnClick, (newUrl) => {
+          let url;
+          if (newUrl) {
+            url = html5TheUrl(newUrl);
+          } else {
+            url = getUrl();
+          }
+          return element.attr('href', url);
```
Comment (sdeering) [inline src/directives/routeOnClick.js]: @lee-borsboom can you check the function above is ok please

File: test/karma/directives/routeOnClick.test.js [hunk 6]
```diff
@@ -22,13 +22,51 @@ describe('routeOnClick directive', function() {
...
+
+      inject(function ($rootScope, $compile, State) {
+        State.set('page', 2);
+        let element = $compile('<a route-on-click="paginationUrlWriter()">Link</a>')($rootScope.$new());
+        $rootScope.$digest();
+        expect(element.attr('href')).toBe('#/page/2');
+
+        $rootScope.paginationUrlWriter = function() {
+          return '/page/3';
+        };
```
Comment (sdeering) [inline test/karma/directives/routeOnClick.test.js]: I'm now not a big fan of the $rootScope stuff, do you think we should upgrade this

File: src/directives/routeOnClick.js (hunk 3 above)
Comment (lee-borsboom) [reply src/directives/routeOnClick.js]: Interesting... was the first condition even working? Maybe go full jQuery on this one:
`if (element.is('a') { ...`

File: test/karma/directives/routeOnClick.test.js [hunk 7]
```diff
@@ -22,10 +22,68 @@ describe('routeOnClick directive', function() {
...
+      inject(function ($rootScope, $compile) {
+        let element = $compile('<a route-on-click="routeLinkUrlWriter()">Link</a>')($rootScope.$new());
+        $rootScope.$digest();
+        expect(element.attr('href')).toBe('#/contacts/update/1');
+      });
+    });
+
+    it('should update the href attribute when the expression eval changes', function() {
+      window.angular.mock.module(function(RouteProvider) {
+        RouteProvider.setPersistentStates('page');
```
Comment (lee-borsboom) [inline test/karma/directives/routeOnClick.test.js]: Do you need RouteProvider.setPersistentStates('page') ?

Comment (sdeering) [reply test/karma/directives/routeOnClick.test.js]: yup it needs it blows up otherwise

**Pull request merged**